woocommerce/woocommerce-ios
# Improve Login Error Handling

Fixes part of #3891.  This improves the error messages that the merchant receives when logging in with a site with an offline device, or an invalid SSL certificate, or a blocked XML-RPC. 

As we discovered in #3891  we default to showing “The website is not a WordPress site...”. That message is not specific and does not help our customers or support folks in troubleshooting further. This PR improves some of that. 

## Design Changes 🎨 

### Offline

This is using the default UI and the default implementation of handling offline messages provided by the [WP Authenticator library](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS). How does this look? 

Before | After 
--------|-------
<img src="https://user-images.githubusercontent.com/198826/114775229-4ebd0900-9d2e-11eb-8db1-525fa9b465e7.gif" width="300">        |       <img src="https://user-images.githubusercontent.com/198826/114775975-2a156100-9d2f-11eb-95b6-0139e002d861.gif" width="300">

### Blocked XML-RPC

Blocking XML-RPC is common. Some hosting providers also block it by default. This UI uses the default provided by the WP Authenticator library. The message is coming from the WP Auth library too. And there [are quite a lot of them](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/4062408fce1596b4744449fd950f9d47ddd4d7eb/WordPressKit/WordPressOrgXMLRPCValidator.swift#L16-L34). Should we display this differently because they're quite technical? 

Before | After 
--------|-------
<img src="https://user-images.githubusercontent.com/198826/114777527-07844780-9d31-11eb-896d-e0ea4ec46428.gif" width="300">|<img src="https://user-images.githubusercontent.com/198826/114777534-09e6a180-9d31-11eb-8ee2-fc5c42c7c0ee.gif" width="300">

### Invalid SSL Certificate

In p91TBi-4Pz-p2, we confirmed that we want to keep the requirement of WooCommerce sites having valid SSL certificates. We currently don't display that information. I added an error message when the merchant tries to log in with a site that does not have SSL. Please suggest better wording! 

Before | After 
--------|-------
<img src="https://user-images.githubusercontent.com/198826/114778298-e5d79000-9d31-11eb-9deb-52645ca7cc14.gif" width="300">|<img src="https://user-images.githubusercontent.com/198826/114778321-eec86180-9d31-11eb-9496-11b6930032cc.gif" width="300">

## Technical Changes ⚙️ 

The WPKit changes in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/382 fixed the problem of WPKit incorrectly returning `NSURLErrorAppTransportSecurityRequiresSecureConnection` (`-1022`) for most unexpected login errors. Now that we have that fixed, the default offline and XML-RPC error messages (shown above) come for free. 🎉 

I then changed the [`AuthenticationError` detection](https://github.com/woocommerce/woocommerce-ios/pull/3957/files#diff-d6518838fddd72ad3679fdd85061c39f0c727d2efec93f40d4a8339c5b7617f8) to preserve the error messages that we currently handle like the `.notValidAddress`. 

I also added a new error screen if the site is not using SSL, [`NoSecureConnectionErrorViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/issue/3891-fix-offline-messaging/WooCommerce/Classes/Authentication/Navigation%20Exceptions/NoSecureConnectionErrorViewModel.swift). 

### Unit Tests

I'm not sure how to add unit tests for these changes. The [`AuthenticationManager.make(with:)`](https://github.com/woocommerce/woocommerce-ios/blob/059da38b74f6581b914f4d42fa4c2cd89682b7cc/WooCommerce/Classes/Authentication/AuthenticationManager.swift#L405) seems like a good candidate but it's just a `switch` so I wasn't sure if it's worth it. Open to suggestions! 

## <a name="testing">Testing</a> 🧪 

1. Log out.
2. Tap “Enter Your Store Address”.
3. Enter a site that does not have HTTPS. If you don't have one, you can use http://httponly.peregrinator.net. The password is available in the Secret Store.
4. Confirm that an error screen is shown: “A secure connection to the site could not be made...”.
5. Go offline.
6. Enter the site URL again.
7. Confirm that the error “The Internet connection appears to be offline.” is displayed. 
8. Go online.
9. Enter a site with blocked XML-RPC. You can use https://do.wpmt.co/ddos. Press Continue.
10. Confirm that the error “... We received a 403 error...” is displayed.
11. Go back to the Login Prologue.
12. Tap “Enter Your Store Address”. 
13. Enter a self-hosted WooCommerce site that has HTTPS. 
14. Confirm that an error is shown: “To use this app, ... you'll need to have the Jetpack plugin installed...”. Or if you activated Jetpack on the site, you should be asked for your WordPress.com password.

## Known Side Effect

There is a bug from WPAuth that is made possible by this change. See https://github.com/woocommerce/woocommerce-ios/issues/4034. I spent some time looking into it but I'm not quite sure whether to fix it in WPAuth or in WC. 

I think it's fine to merge this since the bug only happens when the merchant previously had an XML-RPC error. I'll fix it next. 

## Author Checklist
